cloud/blockstore/libs/endpoints/endpoint_manager.cpp

```diff
@@ -755,13 +763,60 @@ NProto::TStartEndpointResponse TEndpointManager::DoStartEndpoint(
         return promise.ExtractValue();
     }
 
-    auto response = StartEndpointImpl(std::move(ctx), std::move(request), false);
+    auto response =
+        StartEndpointImplWithLock(std::move(ctx), std::move(request), false);
     promise.SetValue(response);
 
     RemoveProcessingSocket(socketPath);
     return response;
 }
 
+// See issue-2841
+// There are two maps Endpoints: one in the TEndpointManager, another in
```

Comment:
I am not completely sure why this mechanism is needed we already have mechanism which synchronizes requests concerning specific socket. This is done using `AddProcessingSocket` and `RemoveProcessingSocket` , which mark/unmark socket as being processes so multiple StartEndpoint can't operate concurrently and the second call will be rejected. There is no need for multi-threading locks/semaphores since this code is always executed from single thread (Executor). If there is some possible race let's add an ut tests to endpoint_manager_ut.cpp so it will be clear what we are fixing and then modify the `AddProcessingSocket` if we want to avoid it